HFE.md

```diff
@@ -0,0 +1,847 @@
+
+```
+HPEXPIREAT key unix-time-milliseconds [NX | XX | GT | LT]
+  FIELDS numfields field [field ...]
+```
+
+`HPEXPIREAT` has the same effect and semantics as `HEXPIREAT``,` but the Unix time at which the field will expire is specified in milliseconds since Unix epoch instead of seconds.
+
+#### HPERSIST
+
+**Synopsis** 
+
+```
+HPERSIST key FIELDS numfields field [field ...]
+```
+
+Remove the existing expiration on a hash key's field(s), turning the field(s) from *volatile* (a field with expiration set) to *persistent* (a field that will never expire as no TTL (time to live) is associated).
+
+#### HSETEX
+
+**Synopsis**
+
+```
+HSETEX key [NX] seconds field value [field value ...]
+```
+
+Similar to `HSET` but adds one or more hash fields that expire after specified number of seconds. By default, this command overwrites the values and expirations of specified fields that exist in the hash. If `NX` option is specified, the field data will not be overwritten. If `key` doesn't exist, a new Hash key is created.
+
+The HSETEX command supports a set of options:
+
+* `NX` — For each specified field, set expiration only when the field has no expiration.
+
+#### HTTL
+
+**Synopsis**
+
+```
+HTTL key FIELDS numfields field [field ...]
+```
+
+Returns the **remaining** TTL (time to live) of a hash key's field(s) that have a set expiration. This introspection capability allows you to check how many seconds a given hash field will continue to be part of the hash key.
+
+#### HPTTL
+
+```
+HPTTL key FIELDS numfields field [field ...]
+```
+
+Like [`HTTL`](https://redis.io/docs/latest/commands/httl/), this command returns the remaining TTL (time to live) of a field that has an expiration set, but in milliseconds instead of seconds.
+
+#### HEXPIRETIME
+
+**Synopsis**
+
+```
+HEXPIRETIME key FIELDS numfields field [field ...]
+```
+
+Returns the absolute Unix timestamp in seconds since Unix epoch at which the given key's field(s) will expire.
+
+#### HPEXPIRETIME
+
+**Synopsis**
+
+```
+HPEXPIRETIME key FIELDS numfields field [field ...]
+```
+
+`HPEXPIRETIME` has the same semantics as `HEXPIRETIME`, but returns the absolute Unix expiration timestamp in milliseconds since Unix epoch instead of seconds.
+
+
+## Keyspace-notification requirements
+
+* A new `hexpire` event will be issued whenever a key field is being assigned with an expiry. This will include both cases of assigning TTL for element which had NO prior TTL assigned and when a TTL is changed on an item which already had TTL assigned. 
+    The same `hexpire` event will be issued for all different commands which manipulate item TTL (e.g.  `HEXPIRE`, `HEXPIREAT`, `HPEXPIRE` etc...)
+    NOTE that for some cases (e.g HSETEX, there will be 2 events issued: `HSET` and `HEXPIRE`)
+* A new `hpersist` event will be issued whenever an item is persisted. this can be when `HPERSIST` was issued.
+* A new `hexpired` event will be issued whenever an item is actually being expired (either actively or lazily) 
```

Comment:
Regarding "being expired" terminology, I think the word expired should always refer to logically expired.

```suggestion
* A new `hexpired` event will be issued whenever an expired item is detected and deleted (either actively or lazily) 
```